### PR TITLE
Add exon information to transcript IDs in Ensembl Controller

### PR DESCRIPTION
**Pull request by dionnezaal**
# What?
This PR adds exon information to the current `EnsemblTranscript` model, called `ExonRange`:
 ```
ExonRange {
exonId (string): Exon id,
exonStart (integer): Start position of exon,
exonEnd (integer): End position of exon ,
rank (integer): Number of exon in transcript ,
strand (integer): Strand exon is on, -1 for - and 1 for +,
version (integer): Exon version
}
```
# Why?
See issue #84 

## This PR includes:
- Added model `ExonRange` to `EnsemblTranscript` Model
- Documentation and transformation script to retrieve exon information from Ensembl and add it to the current transcript information file
- Replacement of script that adds domains to transcript information file (`add_nested_domain_to_ensembl_transcript.py`) with a script that adds both domain and exon information to the transcript information file (`add_domains_and_exon_information_to_ensembl_transcript.py`). The previous script was replaced due to time it took to run. Replacement script can add the information in under a minute

### CODE REVIEW

File: scripts/add_domains_and_exon_information_to_ensembl_transcript.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    # and value is a empty list, unless it is filled by information from file
+    exon_dict = defaultdict(list)
+
+    # Loop through each line of the file with exon information, append information about
+    # exons to dictionary which have transcript id as key
+    with open(exon_file, 'r') as exons:
+        for line in exons:
+            # Split line to retrieve information and
+            list_line = line.split("\t")
+            list_line[6] = list_line[6].replace("\n", "")
```
Comment (sandertan) [inline scripts/add_domains_and_exon_information_to_ensembl_transcript.py]: Please use `strip('\n')` to remove the `\n`

File: scripts/add_domains_and_exon_information_to_ensembl_transcript.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    # This will automatically be empty list when transcript id (key in dict) is requested
+    domain_dict = defaultdict(list)
+
+    # Loop through each line of the file with pfam domain info and append information
+    # to dictionary, which has transcript id as key
+    with open(domains_file, 'r') as domains:
+        for line in domains:
+            # Split line and remove enter from last value
+            list_line = line.split("\t")
+            list_line[6] = list_line[6].replace("\n", "")
```
Comment (sandertan) [inline scripts/add_domains_and_exon_information_to_ensembl_transcript.py]: Please use `strip('\n')` to remove the `\n`, for example:
`            list_line = line.split("\t").strip('\n')`

File: scripts/transform_gff_to_tsv_for_exon_info_from_ensembl.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+    # Dataframe to append the exon information
+    exon_info = pd.DataFrame(columns=['transcriptId', 'exonId', 'exonStart', 'exonEnd', 'rank', 'strand', 'version'])
+    # list to append exon information per row
+    rows = []
+
+    # Open gff file and read lines, when line contains exon information, extract this information
+    with open(gff_file, 'r') as gff:
+        for line in gff:
+            list_line = line.split("\t")
```
Comment (sandertan) [inline scripts/transform_gff_to_tsv_for_exon_info_from_ensembl.py]: Please use `strip('\n')` to remove the `\n`

File: scripts/transform_gff_to_tsv_for_exon_info_from_ensembl.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+                rows.append(data_dict)
+
+    exon_info = exon_info.append(rows, ignore_index=True)
+    return(exon_info)
+
+def main():
+    ## Homo_sapiens.GRCH37.gff3 is local copy from: ftp://ftp.ensembl.org/pub/grch37/release-91/gff3/homo_sapiens/
+    tsv_exon_info = transform_gff_to_tsv("../data/Homo_sapiens.GRCh37.gff3")
+    tsv_exon_info.to_csv(sys.stdout, sep='\t', index=False)
+main()
```
Comment (sandertan) [inline scripts/transform_gff_to_tsv_for_exon_info_from_ensembl.py]: I think this is the proper way to do this. 
```
if __name__ == "__main__":
    main()
```

**Comment by inodb** [comment]
Hi @dionnezaal ! Thanks so much for fixing this and explaining the PR so well. One thing I just discussed with @onursumer is that we want to move the data import stuff to https://github.com/genome-nexus/genome-nexus-importer. I'll try and get this done in the next couple of days. This PR would then prolly be split in two. I can help do that. 

I like that you made `add_domains_and_exon_information_to_ensembl_transcript.py` so much faster. I took a look at why the apply function in pandas was so slow. Adding an index instead of using by column selection  made it way faster. Also noticed this bug: https://github.com/pandas-dev/pandas/issues/9466

The Java part exposing the exons looks good to me!

Might be good to add some tests for the data transformations. Do you have any thoughts on that? A few asserts would already go a long way

**Comment by pieterlukasse** [comment]
@inodb good idea to move the importer parts to a different repo, I had the same thought and had also discussed with @dionnezaal when finishing this PR. However, I think it would be best to try to merge this PR first and then split up the parts in your PR. So, I would like to propose the following: 
- @dionnezaal adds some tests for the python transformation step (Dionne: see unit tests we also have for validation scripts)
- we merge this PR
- @dionnezaal can help with any issues in subsequent PR (where the data loading step is moved to different repo) if necessary

**Comment by dionnezaal** [comment]
@pieterlukasse @inodb I have added a simple unit test for the exon information transformation script which transforms a GFF file with data from two transcripts with exon information and checks if the output file is the same as the example.

**Comment by pieterlukasse** [comment]
Tested the whole solution and works really well :+1:

File: data/Makefile [hunk 5]
```diff
@@ -6,8 +6,11 @@ ensembl_biomart_geneids_grch37.p13.transcript_info.txt: ensembl_biomart_geneids_
 ensembl_biomart_transcripts.txt: ensembl_biomart_geneids_grch37.p13.transcript_info.txt
 	csvcut -tc transcript_stable_id,gene_stable_id,protein_stable_id,protein_length $< | csvsort -c transcript_stable_id,gene_stable_id,protein_stable_id,protein_length | uniq | csvformat -T > $@
 
-ensembl_biomart_transcripts.json: ensembl_biomart_transcripts.txt ensembl_biomart_pfam_grch37.p13.txt
-	cd ../scripts && python add_nested_domain_to_ensembl_transcript.py > ../data/$@
+ensembl_exon_info.txt: Homo_sapiens.GRCh37.gff3.gz
```
Comment (pieterlukasse) [inline data/Makefile]: is `Homo_sapiens.GRCh37.gff3.gz` $1 and can therefore be passed as input to the .py script?

File: data/README.md [hunk 6]
```diff
@@ -28,6 +28,14 @@ Gene stable ID, Transcript stable Id, HGNC Symbol, HGNC ID
 5. Click on `Results`, and export all results to a `TSV` file.
 6. Copy over the downoaded file to replace [ensembl_biomart_geneids_grch37.p13.txt](ensembl_biomart_geneids_grch37.p13.is_canonical.txt).
 
+##### Exon information in Ensembl endpoint
+1. Go to Ensembl FTP site ([ftp://ftp.ensembl.org/pub/grch37/](ftp://ftp.ensembl.org/pub/grch37/))
+2. Click desired Ensembl Release
+3. Click `gff3`
+4. Click `homo_sapiens`
+5. Click Homo_sapiens.GRCh37.[release#].gff3.gz and download will start (currently the file from release 87 is used)
```
Comment (pieterlukasse) [inline data/README.md]: why 87?

File: scripts/transform_gff_to_tsv_for_exon_info_from_ensembl.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
+import pandas as pd
```
Comment (pieterlukasse) [inline scripts/transform_gff_to_tsv_for_exon_info_from_ensembl.py]: please add copyright

File: scripts/unit_test_transformations.py [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+        """Test pfam TSV to internal data structure transformation"""
+        # TODO
+
+    def test_exon_transformation_step(self):
+        """Test exon gff to internal data structure transformation"""
+        # Build up arguments and run
+        out_file_name = 'test_files/exon_information/ensembl_exon_info.txt~'
+        gff_input_file_name = 'test_files/exon_information/sub_Homo_Sapiens.gff3.gz'
+        transform_gff_to_tsv_for_exon_info_from_ensembl.transform_gff_to_tsv(gff_input_file_name, out_file_name)
+        self.assertFileGenerated(out_file_name,'test_files/exon_information/ensembl_exon_info.txt')
```
Comment (pieterlukasse) [inline scripts/unit_test_transformations.py]: very nice :+1: 
However, I get an error when running:
```
AssertionError: 
*** Expected ensembl_exon_info.txt
--- Generated ensembl_exon_info.txt~
***************
*** 1,6 ****
  transcriptId	exonId	exonStart	exonEnd	rank	strand	version
! ENST00000456328	ENSE00002234944	11869	12227	1	1	1
! ENST00000456328	ENSE00003582793	12613	12721	2	1	1
! ENST00000456328	ENSE00002312635	13221	14409	3	1	1
! ENST00000461467	ENSE00001874421	35245	35481	2	-1	1
! ENST00000461467	ENSE00001618781	35721	36073	1	-1	2
--- 1,6 ----
  transcriptId	exonId	exonStart	exonEnd	rank	strand	version
! ENST00000456328	ENSE00002234944	11869	12227	1	1.0	1
! ENST00000456328	ENSE00003582793	12613	12721	2	1.0	1
! ENST00000456328	ENSE00002312635	13221	14409	3	1.0	1
! ENST00000461467	ENSE00001874421	35245	35481	2	-1.0	1
! ENST00000461467	ENSE00001618781	35721	36073	1	-1.0	2


----------------------------------------------------------------------
Ran 2 tests in 0.012s
```

File: data/README.md (hunk 6 above)
Comment (dionnezaal) [reply data/README.md]: Apparantly the GFF file is not update in every release. Therefore, in the latest release [ftp Ensembl release 91](ftp://ftp.ensembl.org/pub/grch37/release-91/gff3/homo_sapiens/) the GFF file is still from release 87.

File: data/Makefile (hunk 5 above)
Comment (dionnezaal) [reply data/Makefile]: Yes, I have implemented this now in the new commits.

File: data/Makefile [hunk 9]
```diff
@@ -6,8 +6,11 @@ ensembl_biomart_geneids_grch37.p13.transcript_info.txt: ensembl_biomart_geneids_
...
+	cd ../scripts && python transform_gff_to_tsv_for_exon_info_from_ensembl.py ../data/$< > ../data/$@
```
Comment (pieterlukasse) [inline data/Makefile]: assuming this last change ` ../data/$<`  was tested

Comment (dionnezaal) [reply data/Makefile]: Yes this was tested

**Comment by inodb** [comment]
Split part of this to genome-nexus-importer https://github.com/genome-nexus/genome-nexus-importer/pull/3

**Comment by inodb** [comment]
Split the rest to #139

**Pull request closed without merging**